Match for generic item in the navigator

hqhhuang opened the pull request:
Bug/issue #, if applicable: 117442044

## Summary
If we are on a specific overload page and it doesn't exist in the navigator, we should attempt to match for its generic item as well.

## Dependencies
Branch: [combine-overloads-navigator](https://github.com/hqhhuang/swift-docc-render/tree/combine-overloads-navigator)

## Testing
Test manually

Steps:
1. Verify if the specific overload page is not curated in the navigator, generic page is highlighted
2. If it is curated, then it should be highlighted as usual. (No behavior change)

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran `npm test`, and it succeeded
- [ ] Updated documentation if necessary

hqhhuang commented:
@swift-ci test

marinaaisa reviewed (approved):
Tested and reviewed. Looking good! Really good job, @hqhhuang :)

dobromir-hristov commented on src/components/Navigator.vue:
```diff
@@ -152,6 +157,21 @@ export default {
+      // eslint-disable-next-line no-param-reassign
+      path = path.replace(/\/$/, '');
+      const hash = isSymbol ? last(path.split('-')) : '';
+      return !!hash.length && hash.length <= 5 && /^[a-z0-9]*$/.test(hash);
```
```suggestion
      return /^[a-z0-9]{1, 5}$/.test(hash);
```

dobromir-hristov commented on src/components/Navigator/NavigatorCard.vue:
```diff
@@ -1014,37 +1018,65 @@ export default {
+          // if no match, try again to match with generic item
+          // Needed for continuing to highlight current generic page
+          // when selecting an overload from dropdown that's also specifically curated in elsewhere
+          const genericItem = lastActivePathItem.split('-')[0];
```
If by any chance we have path like `Foo-bar-f1234` this will use `Foo` which is wrong.

dobromir-hristov commented on src/components/Navigator/NavigatorCard.vue:
```diff
@@ -1014,37 +1018,65 @@ export default {
+        if (last(activePathChildren).path !== lastActivePathItem && this.hasValidHash) {
+          // if item is not found in the tree and its a specific overloaded symbol page
+          // try to match with its generics page instead
+          const genericItem = lastActivePathItem.split('-')[0];
```
Same as above.

dobromir-hristov reviewed (commented):
`hasValidHash` is a bit too generic. I think it should be explicit that its connected to overloads.

hqhhuang commented:
> hasValidHash is a bit too generic. I think it should be explicit that its connected to overloads.
I updated the comment to hopefully make it a bit clearer and changed the name. What do you think?

hqhhuang commented:
@swift-ci test

dobromir-hristov commented on src/components/Navigator/NavigatorCard.vue:
```diff
@@ -1014,37 +1018,65 @@ export default {
+     * try to match if any of the `siblings`,`children` or any of the `parents`
+     * of the UID match the item
+     */
+    matchSurroundingItems(UID, item) {
```
Do you think we should rename this method, so its obvious it mutates the state? I dont really have a better name in mind... 

Also lets rename `item` to `path` as its a path.

hqhhuang replied:
I can add a comment to make it obvious it mutates the state. Can't think of a better name either...

dobromir-hristov commented on tests/unit/components/Navigator.spec.js:
```diff
@@ -248,6 +251,42 @@ describe('Navigator', () => {
+        },
+      },
+    });
+    expect(wrapper.find(NavigatorCard).props('isSpecificOverload')).toBe(true);
```
add a case where the path has no dash or has more than 1 dash.

hqhhuang replied:
The default is no dash. I will add a multiple dash case.

hqhhuang replied:
The test above this tests for no dash/hash. I will add a case for more than 1

dobromir-hristov reviewed (commented):
Looking good. Left few small cleanup comments

hqhhuang commented on src/components/Navigator/NavigatorCard.vue:
```diff
@@ -1014,37 +1018,65 @@ export default {
+          // if no match, try again to match with generic item
+          // Needed for continuing to highlight current generic page
+          // when selecting an overload from dropdown that's also specifically curated in elsewhere
+          const genericItem = lastActivePathItem.slice(0, lastActivePathItem.lastIndexOf('-'));
```
I could, do you think it's worth it? Only 1 line and used in 2 places

hqhhuang commented:
@swift-ci test
